libraries/fc/include/fc/log/trace.hpp

```diff
@@ -1,63 +1,163 @@
+            : id( id )
+            , parent_id( parent_id )
+            , start( time_point::now() )
+            , name( std::move(name) )
+            , log_context( lc )
+            {}
+
+      zipkin_span( const zipkin_span& ) = delete;
+      zipkin_span& operator=( const zipkin_span& ) = delete;
+      zipkin_span& operator=( zipkin_span&& ) = delete;
+
+      zipkin_span( zipkin_span&& rhs ) noexcept
+            : id( rhs.id )
+            , parent_id( rhs.parent_id )
+            , start( rhs.start )
+            , name( std::move( rhs.name ) )
+            , log_context( std::move( rhs.log_context ) )
+            , data( std::move( rhs.data ) )
+            {
+               rhs.id = 0;
+            }
+
+      ~zipkin_span() {
+         if( id == 0 )
+            return;
+         try {
+            auto log = fc::logger::get( zipkin_logger_name );
+            if( log.is_enabled( fc::log_level::debug ) ) {
+               data["zipkin.id"] = id;
+               static_assert( std::is_same<decltype(start.time_since_epoch().count()), int64_t>::value, "Expect to store int64_t");
+               data["zipkin.timestamp"] = start.time_since_epoch().count();
+               static_assert( std::is_same<decltype((time_point::now() - start).count()), int64_t>::value, "Expect to store int64_t");
+               data["zipkin.duration"] = (time_point::now() - start).count();
+               data["zipkin.name"] = std::move(name);
+               if( parent_id != 0 ) {
+                  data["zipkin.traceId"] = parent_id;
+                  data["zipkin.parentId"] = parent_id;
+               } else {
+                  data["zipkin.traceId"] = id;
+               }
+               log.log( fc::log_message( log_context, "", std::move( data ) ) );
+            }
+         } catch( ... ) {}
+      }
+
+      template<typename T>
+      void add_tag( const std::string& key, T&& var ) {
+         data( key, std::forward<T>( var ) );
+      }
 
-/// Simple wrappers for cppkin tracing
+      struct token {
+         friend zipkin_span;
+         friend struct zipkin_trace;
+         friend struct optional_trace;
+         constexpr explicit operator bool() const noexcept { return id != 0; }
+      private:
+         explicit token(uint64_t id)
+         : id(id) {}
+         uint64_t id;
+      };
 
-inline bool fc_cppkin_trace_enabled = false;
+      token get_token() const { return token{id}; };
 
+      static uint64_t to_id( const fc::sha256& id ) {
+         return id._hash[3];
+      }
+
+      template<typename T>
+      static uint64_t to_id( const T& id ) {
+         static_assert( std::is_same_v<decltype(id.data()), const uint64_t*>, "expected uint64_t" );
+         return id.data()[3];
+      }
+
+   private:
+      uint64_t id;
+      // zipkin traceId and parentId are same since only allowing trace with span children, no children of spans
```

Comment:
My understanding is "root span" <==> "trace". So one span (the root span, or trace) has no parentId. All the spans then share the traceId. Since currently only supporting trace with spans (no child spans of spans) then traceId and parentId are always the same.

Reply:
Updated the comment to be more clear.